arch/src/aarch64/gic.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{io, result};
+
+use kvm::{DeviceFd, VmFd};
+
+// Unfortunately bindgen omits defines that are based on other defines.
+// See arch/arm64/include/uapi/asm/kvm.h file from the linux kernel.
+const SZ_64K: u64 = 0x0001_0000;
+const KVM_VGIC_V3_DIST_SIZE: u64 = SZ_64K;
+const KVM_VGIC_V3_REDIST_SIZE: u64 = (2 * SZ_64K);
+
+#[derive(Debug)]
+pub enum Error {
+    /// Error while calling KVM ioctl for setting up the global interrupt controller.
+    CreateGIC(io::Error),
+    /// Error while setting device attributes for the GIC.
+    SetDeviceAttribute(io::Error),
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+/// Create a GICv3 device.
+///
+/// Logic from this function is based on virt/kvm/arm/vgic/vgic-kvm-device.c from linux kernel.
+pub fn create_gicv3(vm: &VmFd, vcpu_count: u8) -> Result<DeviceFd> {
+    /* We are creating a V3 GIC.
+     As per https://static.docs.arm.com/dai0492/b/GICv3_Software_Overview_Official_Release_B.pdf,
+     section 3.5 Programmers' model, the register interface of a GICv3 interrupt controller is split
+     into three groups: distributor, redistributor, CPU.
+     As per Figure 9 from same section, there is 1 Distributor and multiple redistributors (one per
+     each CPU).
+    */
+    let mut gic_device = kvm_bindings::kvm_create_device {
+        type_: kvm_bindings::kvm_device_type_KVM_DEV_TYPE_ARM_VGIC_V3,
+        fd: 0,
+        flags: 0,
+    };
+
+    let vgic_fd = vm
+        .create_device(&mut gic_device)
+        .map_err(Error::CreateGIC)?;
+
+    /* Setting up the distributor attribute.
+     We are placing the GIC below 1GB so we need to substract the size of the distributor.
+    */
+    let dist_if_addr: u64 = super::layout::MAPPED_IO_START - KVM_VGIC_V3_DIST_SIZE;
```

Comment:
I'm not sure what the story is here. Does the guest memory have a gap like in x86? What is the guest address space region being used for devices (like this GIC, as well as following mmio devices)?

Reply:
Take a look in [layout file](https://github.com/firecracker-microvm/firecracker/blob/master/arch/src/aarch64/layout.rs) for info on the layout of the memory. Indeed, the GIC will reside between 0 and 1GB and the mmio devices reside between 1gb and 2gb. The DRAM will start from 2Gb. The gap from x86 is a hole inside the DRAM itself. We are not splitting DRAM in aarch64. But yes, we have holes early in the guest physical address space reserved for devices.